Added a sensor to monitor event on repositories in the BitBucket Server/Cloud.

userlocalhost opened the pull request:
I implemented a sensor to monitor the new commits for each repositories in the  BitBucket cloud/server(for #2,#3).

Thank you.

userlocalhost commented:
Sorry, I didn't take account of #4.

I fixed `pack.yml` and added description for this sensor.

Thank you.

Kami commented on config.schema.yaml:
```diff
@@ -14,3 +14,21 @@
     secret: false
     required: true
+  sensor:
+    type: "object"
```
Need to add `additionalProperties: false` attribute there otherwise user will be able to provider arbitrary invalid options as part of setup key.

For example:

```yaml
setup:
  foo: 'bar'
  bar: 'baz'
```

This would all work if `additionalProperties: false` is not specified.

Kami commented on config.schema.yaml:
```diff
@@ -14,3 +14,21 @@
+        description: "URL of BitBucket server (this is only required to set in the case of the 'bitbucket_type' is set 'server')"
+        required: false
+      repositories:
+        type: "array"
```
Should also add:

```yaml
items:
  type: "string"
```

This tells it that each array item needs to be a string.

Kami commented on sensors/repository_sensor.py:
```diff
@@ -0,0 +1,186 @@
+            repo_name = '%s/%s' % (proj, repo)
+            for branch in self.client.projects[proj].repos[repo].branches():
+                for commit in self.client.projects[proj].repos[repo].commits(branch['id']):
+                    if self.last_commit[repo_name][branch['displayId']] == commit['id']:
```
Just thinking out loud and trying to understand if this way of detecting last / new commits will work in all scenarios.

I guess it won't work correctly if user force pushes a change, but there is not much we can do in that case anyway.

userlocalhost replied:
Thank you for your comment.
To deal with the case that git commit tree is changed, I'll changed to compare the commit using committed time.

Kami replied:
Yeah, I think committed time is better.

It can still be changed with a force push, but I think it's still more robust.

userlocalhost replied:
Thank you again.

You're right. An amended commit can't be notified with the time comparing because that will be same time with the replaced one, but different to hash value.

I want to make it shelve because it needs consideration to solve, I guess.
After refactoring to the time-based comparing processing, I'll make na issue for it.

Kami commented on sensors/repository_sensor.py:
```diff
@@ -0,0 +1,186 @@
+        for (proj, repo) in [x.split('/') for x in self.repositories]:
+            repo_name = '%s/%s' % (proj, repo)
+            for branch in self.client.projects[proj].repos[repo].branches():
+                for commit in self.client.projects[proj].repos[repo].commits(branch['id']):
```
How does this work if there are many commits?

Does `commits()` return some kind of generator or it only returns up to X last commits?

In any case, if possible, it would be great if we could limit number of recent commits returned here or similar (e.g. only return new commits since TIMESTAMP).

Otherwise it could get quite expensive and slow constantly polling and iterating over 1000s of commits during each poll interval.

Kami commented:
Nice work, thanks.

Overall the change looks good to me, but there are some small things which can be improved and some things which could potentially be optimized / I need a bit more clarification on.

userlocalhost commented:
Thank you very much for your review!
I understood all comments that you gave, and I should be able to correct them.

userlocalhost commented:
I've corrected the things that are commented.
On an assumption to handle a huge number of commits, I've added a feature to set timeout.

But current implementation leave the problem for handling amended commit (reported in #6).
This should be able to be corrected.

Thank you.

Kami commented:
Thanks, will review it shortly.

Kami commented on README.md:
```diff
@@ -195,6 +195,31 @@ Usage:
+
+Currently, this supports following event type.
+
+* `commit` - Triggered when new commit(a) are made.
```
`(s)` :)

userlocalhost replied:
Sorry, I'll correct it soon.

Kami commented on sensors/repository_sensor.py:
```diff
@@ -0,0 +1,221 @@
+    def poll(self):
+        @timeout(self.TIMEOUT_SECONDS)
+        def get_server_updated_commits():
+            for (proj, repo) in [x.split('/') for x in self.repositories]:
```
This would be a bit more readable if there were less levels of nesting (split it in multiple methods), but it's fine for now.

Kami commented on sensors/repository_sensor.py:
```diff
@@ -0,0 +1,221 @@
+            commit_time = datetime.strptime(commit['time'], self.DATE_FORMAT)
+
+            if self.last_commit[commit['repository']][commit['branch']] < commit_time:
+                self.last_commit[commit['repository']][commit['branch']] = commit_time
```
To prevent same commits to be replayed when sensor restarts you should also store this value in datastore after each `poll` loop and retrieve it in setup on initialization when server starts up.

You can find examples on how we do that in some other sensors (e.g. Twitter).

userlocalhost replied:
Thank you for your advice.
I was lack of the care about the possibility of duplicate dispatching in the huge update case.

I'll consider to prevent dispatching commit which will be dispatched in another polling processing with reference to the implementation of Twitter Sensor.

Kami commented:
There are still some things which can be improved, but it's good enough for now so I will go ahead and merge it.

Thanks again.
